Sources/GRPC/FakeResponseStream.swift

```diff
@@ -0,0 +1,321 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import NIO
+import NIOHPACK
+
+/// A fake response stream into which users may inject response parts for use in unit tests.
+///
+/// Users may not interact with this class directly but may do so via one of its subclasses
+/// `FakeUnaryResponse` and `FakeStreamingResponse`.
+public class FakeResponseStream<Request: GRPCPayload, Response: GRPCPayload> {
+  private enum StreamEvent {
+    case responsePart(_GRPCClientResponsePart<Response>)
+    case error(Error)
+  }
+
+  /// The channel to use for communication.
+  internal let channel: EmbeddedChannel
+
+  /// A buffer to hold responses in before the proxy is activated.
+  private var responseBuffer: CircularBuffer<StreamEvent>
+
+  /// The current state of the proxy.
+  private var activeState: ActiveState
+
+  /// The state of sending response parts.
+  private var sendState: SendState
+
+  private enum ActiveState {
+    case inactive
+    case active
+  }
+
+  private enum SendState {
+    // Nothing has been sent; we can send initial metadata to become 'sending' or trailing metadata
+    // to start 'closing'.
+    case idle
+
+    // We're sending messages. We can send more messages in this state or trailing metadata to
+    // transition to 'closing'.
+    case sending
+
+    // We're closing: we've sent trailing metadata, we may only send a status now to close.
+    case closing
+
+    // Closed, nothing more can be sent.
+    case closed
+  }
+
+  internal init() {
+    self.activeState = .inactive
+    self.sendState = .idle
+    self.responseBuffer = CircularBuffer()
+    self.channel = EmbeddedChannel()
+  }
+
+  /// Activate the test proxy; this should be called
+  internal func activate() {
+    switch self.activeState {
+    case .inactive:
+      // Activate the channel. This will allow any request parts to be sent.
+      self.channel.pipeline.fireChannelActive()
+
+      // Unbuffer any response parts.
+      while !self.responseBuffer.isEmpty {
+        self.write(self.responseBuffer.removeFirst())
+      }
+
+      // Now we're active.
+      self.activeState = .active
+
+    case .active:
+      ()
+    }
+  }
+
+  /// Write or buffer the response part, depending on the our current state.
+  internal func _sendResponsePart(_ part: _GRPCClientResponsePart<Response>) {
+    self.send(.responsePart(part))
+  }
+
+  internal func _sendError(_ error: Error) {
+    self.send(.error(error))
+  }
+
+  private func send(_ part: StreamEvent) {
+    switch self.filter(part) {
+    case .include:
+      self.writeOrBuffer(part)
+
+    case .includeAndPrepend(let extraPart):
+      self.writeOrBuffer(extraPart)
+      self.writeOrBuffer(part)
+
+    case .exclude:
+      ()
```

Comment:
What's the reasoning for tolerating these errors? Should the user not deal with the crash, given that this is a test helper?

Reply:
It's marginally easier to use in the happy case – which isn't exactly a strong reason – I'll change this.